File: src/org/ggp/base/util/prover/aima/substituter/Substituter.java
package org.ggp.base.util.prover.aima.substituter

import java.util.ArrayList
import java.util.List

import org.ggp.base.util.gdl.grammar.GdlConstant
import org.ggp.base.util.gdl.grammar.GdlDistinct
import org.ggp.base.util.gdl.grammar.GdlFunction
import org.ggp.base.util.gdl.grammar.GdlLiteral
import org.ggp.base.util.gdl.grammar.GdlNot
import org.ggp.base.util.gdl.grammar.GdlOr
import org.ggp.base.util.gdl.grammar.GdlPool
import org.ggp.base.util.gdl.grammar.GdlProposition
import org.ggp.base.util.gdl.grammar.GdlRelation
import org.ggp.base.util.gdl.grammar.GdlRule
import org.ggp.base.util.gdl.grammar.GdlSentence
import org.ggp.base.util.gdl.grammar.GdlTerm
import org.ggp.base.util.gdl.grammar.GdlVariable
import org.ggp.base.util.prover.aima.substitution.Substitution


class Substituter


    def static GdlLiteral substitute(GdlLiteral literal, Substitution theta)
	
        return substituteLiteral(literal, theta)

    def static GdlSentence substitute(GdlSentence sentence, Substitution theta)
	
        return substituteSentence(sentence, theta)

    def static GdlRule substitute(GdlRule rule, Substitution theta)
	
        return substituteRule(rule, theta)

    def GdlConstant substituteConstant(GdlConstant constant, Substitution theta)
	
        return constant

    def GdlDistinct substituteDistinct(GdlDistinct distinct, Substitution theta)
	
        if (distinct.isGround())
		
            return distinct
        else
		
            GdlTerm arg1 = substituteTerm(distinct.getArg1(), theta)
            GdlTerm arg2 = substituteTerm(distinct.getArg2(), theta)

            return GdlPool.getDistinct(arg1, arg2)

    def GdlFunction substituteFunction(GdlFunction function, Substitution theta)
	
        if (function.isGround())
		
            return function
        else
		
            GdlConstant name = substituteConstant(function.getName(), theta)

            List<GdlTerm> body = new ArrayList<GdlTerm>()
            for (int i = 0; i < function.arity(); i++)
			
                body.add(substituteTerm(function.get(i), theta))

            return GdlPool.getFunction(name, body)

    def GdlLiteral substituteLiteral(GdlLiteral literal, Substitution theta)
	
        if (literal instanceof GdlDistinct)
		
            return substituteDistinct((GdlDistinct) literal, theta)
        else if (literal instanceof GdlNot)
		
            return substituteNot((GdlNot) literal, theta)
        else if (literal instanceof GdlOr)
		
            return substituteOr((GdlOr) literal, theta)
        else
		
            return substituteSentence((GdlSentence) literal, theta)

    def GdlNot substituteNot(GdlNot not, Substitution theta)
	
        if (not.isGround())
		
            return not
        else
		
            GdlLiteral body = substituteLiteral(not.getBody(), theta)
            return GdlPool.getNot(body)

    def GdlOr substituteOr(GdlOr or, Substitution theta)
	
        if (or.isGround())
		
            return or
        else
		
            List<GdlLiteral> disjuncts = new ArrayList<GdlLiteral>()
            for (int i = 0; i < or.arity(); i++)
			
                disjuncts.add(substituteLiteral(or.get(i), theta))

            return GdlPool.getOr(disjuncts)

    def GdlProposition substituteProposition(GdlProposition proposition, Substitution theta)
	
        return proposition

    def GdlRelation substituteRelation(GdlRelation relation, Substitution theta)
	
        if (relation.isGround())
		
            return relation
        else
		
            GdlConstant name = substituteConstant(relation.getName(), theta)

            List<GdlTerm> body = new ArrayList<GdlTerm>()
            for (int i = 0; i < relation.arity(); i++)
			
                body.add(substituteTerm(relation.get(i), theta))

            return GdlPool.getRelation(name, body)

    def GdlSentence substituteSentence(GdlSentence sentence, Substitution theta)
	
        if (sentence instanceof GdlProposition)
		
            return substituteProposition((GdlProposition) sentence, theta)
        else
		
            return substituteRelation((GdlRelation) sentence, theta)

    def GdlTerm substituteTerm(GdlTerm term, Substitution theta)
	
        if (term instanceof GdlConstant)
		
            return substituteConstant((GdlConstant) term, theta)
        else if (term instanceof GdlVariable)
		
            return substituteVariable((GdlVariable) term, theta)
        else
		
            return substituteFunction((GdlFunction) term, theta)

    def GdlTerm substituteVariable(GdlVariable variable, Substitution theta)
	
        if (!theta.contains(variable))
		
            return variable
        else
		
            GdlTerm result = theta.get(variable)
            GdlTerm betterResult = null

            while (!(betterResult = substituteTerm(result, theta)).equals(result))
			
                result = betterResult

            theta.put(variable, result)
            return result

    def GdlRule substituteRule(GdlRule rule, Substitution theta)
	
        GdlSentence head = substitute(rule.getHead(), theta)

        List<GdlLiteral> body = new ArrayList<GdlLiteral>()
        for ( GdlLiteral literal : rule.getBody() )
		
            body.add(substituteLiteral(literal, theta))

        return GdlPool.getRule(head, body)

File: ggpy/cruft/autocode/Substituter.py
#!/usr/bin/env python
""" generated source for module Substituter """
# package: org.ggp.base.util.prover.aima.substituter
import java.util.ArrayList

import java.util.List

import org.ggp.base.util.gdl.grammar.GdlConstant

import org.ggp.base.util.gdl.grammar.GdlDistinct

import org.ggp.base.util.gdl.grammar.GdlFunction

import org.ggp.base.util.gdl.grammar.GdlLiteral

import org.ggp.base.util.gdl.grammar.GdlNot

import org.ggp.base.util.gdl.grammar.GdlOr

import org.ggp.base.util.gdl.grammar.GdlPool

import org.ggp.base.util.gdl.grammar.GdlProposition

import org.ggp.base.util.gdl.grammar.GdlRelation

import org.ggp.base.util.gdl.grammar.GdlRule

import org.ggp.base.util.gdl.grammar.GdlSentence

import org.ggp.base.util.gdl.grammar.GdlTerm

import org.ggp.base.util.gdl.grammar.GdlVariable

import org.ggp.base.util.prover.aima.substitution.Substitution

class Substituter(object):
    """ generated source for class Substituter """
    @classmethod
    @overloaded
    def substitute(cls, literal, theta):
        """ generated source for method substitute """
        return substituteLiteral(literal, theta)

    @classmethod
    @substitute.register(object, GdlSentence, Substitution)
    def substitute_0(cls, sentence, theta):
        """ generated source for method substitute_0 """
        return substituteSentence(sentence, theta)

    @classmethod
    @substitute.register(object, GdlRule, Substitution)
    def substitute_1(cls, rule, theta):
        """ generated source for method substitute_1 """
        return substituteRule(rule, theta)

    @classmethod
    def substituteConstant(cls, constant, theta):
        """ generated source for method substituteConstant """
        return constant

    @classmethod
    def substituteDistinct(cls, distinct, theta):
        """ generated source for method substituteDistinct """
        if distinct.isGround():
            return distinct
        else:
            return GdlPool.getDistinct(arg1, arg2)

    @classmethod
    def substituteFunction(cls, function_, theta):
        """ generated source for method substituteFunction """
        if function_.isGround():
            return function_
        else:
            while i < function_.arity():
                body.add(substituteTerm(function_.get(i), theta))
                i += 1
            return GdlPool.getFunction(name, body)

    @classmethod
    def substituteLiteral(cls, literal, theta):
        """ generated source for method substituteLiteral """
        if isinstance(literal, (GdlDistinct, )):
            return cls.substituteDistinct(literal, theta)
        elif isinstance(literal, (GdlNot, )):
            return substituteNot(literal, theta)
        elif isinstance(literal, (GdlOr, )):
            return substituteOr(literal, theta)
        else:
            return substituteSentence(literal, theta)

    @classmethod
    def substituteNot(cls, not_, theta):
        """ generated source for method substituteNot """
        if not_.isGround():
            return not_
        else:
            return GdlPool.getNot(body)

    @classmethod
    def substituteOr(cls, or_, theta):
        """ generated source for method substituteOr """
        if or_.isGround():
            return or_
        else:
            while i < or_.arity():
                disjuncts.add(cls.substituteLiteral(or_.get(i), theta))
                i += 1
            return GdlPool.getOr(disjuncts)

    @classmethod
    def substituteProposition(cls, proposition, theta):
        """ generated source for method substituteProposition """
        return proposition

    @classmethod
    def substituteRelation(cls, relation, theta):
        """ generated source for method substituteRelation """
        if relation.isGround():
            return relation
        else:
            while i < relation.arity():
                body.add(substituteTerm(relation.get(i), theta))
                i += 1
            return GdlPool.getRelation(name, body)

    @classmethod
    def substituteSentence(cls, sentence, theta):
        """ generated source for method substituteSentence """
        if isinstance(sentence, (GdlProposition, )):
            return cls.substituteProposition(sentence, theta)
        else:
            return cls.substituteRelation(sentence, theta)

    @classmethod
    def substituteTerm(cls, term, theta):
        """ generated source for method substituteTerm """
        if isinstance(term, (GdlConstant, )):
            return cls.substituteConstant(term, theta)
        elif isinstance(term, (GdlVariable, )):
            return substituteVariable(term, theta)
        else:
            return cls.substituteFunction(term, theta)

    @classmethod
    def substituteVariable(cls, variable, theta):
        """ generated source for method substituteVariable """
        if not theta.contains(variable):
            return variable
        else:
            while not (betterResult = cls.substituteTerm(result, theta)) == result:
                result = betterResult
            theta.put(variable, result)
            return result

    @classmethod
    def substituteRule(cls, rule, theta):
        """ generated source for method substituteRule """
        head = cls.substitute(rule.getHead(), theta)
        body = ArrayList()
        for literal in rule.getBody():
            body.add(cls.substituteLiteral(literal, theta))
        return GdlPool.getRule(head, body)

